src/utils/parser/getScopes.js

```diff
@@ -0,0 +1,280 @@
+        const scope = _createTempScope("Function", parent, location);
+        if (isNode(tree.id, "Identifier")) {
+          const functionName = { kind: "fn", refs: [] };
+          _getFunctionScope(parent).names[tree.id.name] = functionName;
+          scope.names[tree.id.name] = functionName;
+        }
+        tree.params.forEach(param => {
+          _parseDeclarator(param, scope, "param");
+        });
+        parent = scope;
+        return;
+      }
+      if (path.isForInStatement() || path.isForStatement()) {
+        const init = tree.init || tree.left;
+        let scope = parent;
+        if (
+          isNode(init, "VariableDeclaration") &&
+          (init.kind === "let" || init.kind === "const")
+        ) {
+          savedParents.set(path, parent);
+          scope = _createTempScope("For", parent, location);
+          parent = scope;
+        }
+        return;
+      }
+      if (path.isCatchClause()) {
+        savedParents.set(path, parent);
+        const scope = _createTempScope("Catch", parent, location);
+        _parseDeclarator(tree.param, scope, "param");
+        parent = scope;
+        return;
+      }
+      if (path.isBlockStatement()) {
+        const hasLet = tree.body.some(node => {
+          if (!isNode(node, "VariableDeclaration")) {
+            return false;
+          }
+          return node.kind === "let" || node.kind === "const";
+        });
+        let scope = parent;
+        if (hasLet) {
+          savedParents.set(path, parent);
+          scope = _createTempScope("Block", parent, location);
+          parent = scope;
+        }
+        return;
+      }
+      if (path.isVariableDeclaration()) {
+        let hoistAt = parent;
+        if (tree.kind !== "let" && tree.kind !== "const") {
+          hoistAt = _getFunctionScope(parent);
+        }
+        tree.declarations.forEach(declarator => {
+          _parseDeclarator(declarator.id, hoistAt, tree.kind);
+        });
+        return;
+      }
+      if (path.isIdentifier()) {
+        _recordIdentifierPosition(parent, tree.name, tree.loc);
+      }
+    },
+    exit(path: NodePath) {
+      const savedParent = savedParents.get(path);
+      if (savedParent) {
+        parent = savedParent;
+        savedParents.delete(path);
+      }
+    }
+  });
+  return _toGeneratedScopes(parent.scopes, source.id);
+}
+
+function _compareLocations(a: Location, b: Location): number {
+  // $FlowIgnore
+  return a.line == b.line ? a.column - b.column : a.line - b.line;
+}
+
+var parsedScopesCache = new Map();
+
+export default function getScopes(location: Location): SourceScope[] {
+  const { sourceId } = location;
+  let parsedScopes = parsedScopesCache.get(sourceId);
+  if (!parsedScopes) {
+    parsedScopes = _parseJSScopes(getSource(sourceId));
+    parsedScopesCache.set(sourceId, parsedScopes);
+  }
+
+  // Find inner most
+  let search = parsedScopes;
+  let found = [];
+  while (search) {
+    let foundOne = search.some(s => {
```

Comment:
search.some stops execution after one is found; but we are modifying the found array here and the search object. perhaps what we want here again is reduce?

```javascript
const found = search.reduce((found, s) => {
  if (
    _compareLocations(s.start, location) <= 0 &&
    _compareLocations(location, s.end) < 0
  ) {
    return [s, ...found];
  }
  return found;
})
```

you can also factor that out into an independant function and do (imagining that the reduce is its own function, and the map is another):
```javascript
return search.reduce(findLocations).map(createScope);
```